Recipe: Expanding arrays into new rows (#53)
ververica/flink-sql-cookbook: other-builtin-functions/05_exploding_arrays/05_exploding_arrays.md

# 04 Expanding arrays into new rows

> ![Twitter Badge](https://img.shields.io/badge/Flink%20Version-1.3%2B-lightgrey)

Reviewer: Flink 1.3?

Author: I've checked all the documentation since then and at least the docs say it was supported :)

The source table (`HarryPotter`) is backed by the [`faker` connector](https://flink-packages.org/packages/flink-faker), which continuously generates rows in memory based on Java Faker expressions.

There are many cases where data contains complex data types. Complex data types means that the data is structured in a nested way. Examples of these data types are `ARRAY`, `MAP` or `ROW`.

In this recipe, we'll expand an `ARRAY` into a new row for each element using `CROSS JOIN UNNEST` and join the new rows with a lookup table.

This table DDL creates a `HarryPotter` themed table. It contains a character from Harry Potter and 3 spells that the character used. The `spells` is of data type `ARRAY`.

## Script

```sql
-- Create source table
CREATE TABLE `HarryPotter` (
  `character` STRING,
  `spells` ARRAY<STRING>
) WITH (
  'connector' = 'faker',
  'fields.character.expression' = '#{harry_potter.character}',
  'fields.spells.expression' = '#{harry_potter.spell}',
  'fields.spells.length' = '3'
);
```

When querying this table, your results will look like this:

![05_complex_data_types](05_complex_data_types.png)

> In order to generate new rows for each element in the `spells` array, we’ll use a `CROSS JOIN UNNEST`. By applying this statement, the `UNNEST` will first create new rows for each element, which are stored in a temporary table `SpellsTable`. Secondly, the `CROSS JOIN` joins the newly generated rows from `SpellsTable` with the matching `HarryPotter` data. The DDL below will create a view that we’ll use to join all the newly generated rows with a lookup table.

Reviewer: ```suggestion
In order to generate new rows for each element in the `spells` array, we’ll use a `CROSS JOIN UNNEST`. By applying this statement, the `UNNEST` will create one row for each element in `spells`, which we will store in a temporary table `SpellsTable`. Secondly, the `CROSS JOIN` joins each row in the `SpellsTable` with the matching row  of the `HarryPotter` table. The DDL below will create a view that we’ll use to join all the newly generated rows with a lookup table.
```

```sql
CREATE TEMPORARY VIEW `SpellsPerCharacter` AS
  SELECT `HarryPotter`.`character`, `SpellsTable`.`spell`

>   FROM HarryPotter CROSS JOIN UNNEST(HarryPotter.spells) AS SpellsTable (spell);

Reviewer: line break before CROSS JOIN for consistency?

```

The results of this view will contain each individual spell as a row. The results look like the following:

![05_unnested_data](05_unnested_data.png)

> You can then use these results like you normally would, for example for joining with a lookup table. In this example, you would join the `spell` with the `spoken_language` so that you could see in which language the spells were spoken by the characters.

Reviewer: ```suggestion
You can then use the resulting table like you normally would, for example, for joining with a lookup table. Here, we enrich each `spell` with the `spoken_language` so that you can see in which language the spells were cast.
```

# 04 Expanding arrays into new rows

![Twitter Badge](https://img.shields.io/badge/Flink%20Version-1.3%2B-lightgrey)

> :bulb: This example will show how to create new rows for each element in an array using a `CROSS JOIN UNNEST`.

The source table (`HarryPotter`) is backed by the [`faker` connector](https://flink-packages.org/packages/flink-faker), which continuously generates rows in memory based on Java Faker expressions.

There are many cases where data contains complex data types. Complex data types means that the data is structured in a nested way. Examples of these data types are `ARRAY`, `MAP` or `ROW`.

In this recipe, we'll expand an `ARRAY` into a new row for each element using `CROSS JOIN UNNEST` and join the new rows with a lookup table.

> This table DDL creates a `HarryPotter` themed table with three multiple complex data types. It contains a `character_with_age` of data type `MAP<STRING, INT>` that contains three Harry Potter characters and their age. Also included is `spells` of data type `ARRAY` that contain three spells, spoken by them. The last included complex data type is `house_points` of data type ``ROW<`house` STRING, `points` INT>`` that refers to the house of characters and the points given for that spell.

Reviewer: ```suggestion
This table DDL creates a `HarryPotter` themed table with three complex data types. It contains a `character_with_age` of data type `MAP<STRING, INT>` that contains three Harry Potter characters and their age. Also included is `spells` of data type `ARRAY` that contain three spells, spoken by them. The last included complex data type is `house_points` of data type ``ROW<`house` STRING, `points` INT>`` that refers to the house of characters and the points given for that spell.
```

## Script

```sql
-- Create source table

> CREATE TABLE `HarryPotter` (

Reviewer: What do you think about making this table simple by removing the MAP and ROW columns? They don't seems to be relevant for the recipe and distract/make it more complicated to follow along.

Author: Agreed, I'll use those in different recipes in the future
